### PR TITLE
7172359: HTML parser StackOverflowError on invalid HTML: <li> tag inside an <option>

### PR DESCRIPTION
**Pull request by prsadhuk**
If there is invalid tag, stack terminates but it still tries to recurse through by calling legalElementContext() which results in StackOverflowError.
Fix is to return if stack is terminated after checking current tag is not valid in current context

All other test/html/parser as well as other jtreg suite tests are ok with the fix.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-7172359](https://bugs.openjdk.org/browse/JDK-7172359): HTML parser StackOverflowError on invalid HTML: <li> tag inside an <option>

### Reviewers
 * [Tejesh R](https://openjdk.org/census#tr) (@TejeshR13 - Author) ⚠️ Review applies to [4296729f](https://git.openjdk.org/jdk/pull/10279/files/4296729f7fd28f77b5a4a8a09244bbc221f7c744)
 * [Jayathirth D V](https://openjdk.org/census#jdv) (@jayathirthrao - **Reviewer**) ⚠️ Review applies to [4296729f](https://git.openjdk.org/jdk/pull/10279/files/4296729f7fd28f77b5a4a8a09244bbc221f7c744)
 * [Alexey Ivanov](https://openjdk.org/census#aivanov) (@aivanov-jdk - **Reviewer**) ⚠️ Review applies to [46c477d8](https://git.openjdk.org/jdk/pull/10279/files/46c477d84f36454d4988a40da3ce77437bc16241)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10279/head:pull/10279` \
`$ git checkout pull/10279`

Update a local copy of the PR: \
`$ git checkout pull/10279` \
`$ git pull https://git.openjdk.org/jdk pull/10279/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10279`

View PR using the GUI difftool: \
`$ git pr show -t 10279`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10279.diff">https://git.openjdk.org/jdk/pull/10279.diff</a>

</details>

### CODE REVIEW

**Review by TejeshR13 — APPROVED** [review]
Fix is tested and works fine.

**Review by jayathirthrao — APPROVED** [review]

File: test/jdk/javax/swing/text/html/parser/ParserCrash.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+ * @summary  Verifies HTML parser StackOverflowError on invalid HTML: <li> tag
+ *           inside an <option>
+ * @run main ParserCrash
+ */
+import javax.swing.text.html.HTMLDocument;
+import javax.swing.text.html.HTMLEditorKit;
+import java.io.StringReader;
+import java.io.Reader;
+
+public class ParserCrash {
```
[the remark this answers is not in the document]
Comment (prsadhuk) [reply test/jdk/javax/swing/text/html/parser/ParserCrash.java]: ok

File: test/jdk/javax/swing/text/html/parser/ParserStackOverflow.java [hunk 2]
```diff
@@ -45,7 +45,7 @@ public static void main( String[] argv ) throws Exception {
         Reader reader = new StringReader(badHtml);
         kit.read(reader, doc, 0); // StackOverflowError here
 
-        System.out.println("Succeeded! (expected StackOverflowError");
+        System.out.println("Succeeded! (no StackOverflowError");
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/parser/ParserStackOverflow.java:48]:
```suggestion
        System.out.println("Succeeded! (no StackOverflowError)");
```
The closing parenthesis is missing in the string.

File: test/jdk/javax/swing/text/html/parser/ParserStackOverflow.java [hunk 3]
```diff
@@ -25,14 +25,14 @@
...
  */
 import javax.swing.text.html.HTMLDocument;
 import javax.swing.text.html.HTMLEditorKit;
 import java.io.StringReader;
 import java.io.Reader;
 
-public class ParserCrash {
+public class ParserStackOverflow {
     /*
      * li element inside an option will crash javax.swing.text.html.parser.Parser
```
Comment (aivanov-jdk) [inline test/jdk/javax/swing/text/html/parser/ParserStackOverflow.java:37]: It doesn't **crash** it.

**Review by aivanov-jdk — APPROVED** [review]

**Pull request closed without merging**